Opt-in layer caching (#209)
buildpacks/spec: buildpack.md

Using the [Layer Content Metadata](#layer-content-metadata-toml) provided by a buildpack in a `<layers>/<layer>.toml` file, the lifecycle MUST determine:

- Whether the layer directory `<layers>/<layer>/` should be available to the app (via the `launch` boolean).
- Whether the layer directory `<layers>/<layer>/` should be available to subsequent buildpacks (via the `build` boolean).
- Whether the layer directory `<layers>/<layer>/` should be persisted and made available to subsequent builds of the same OCI image (via the `cache` boolean).

All combinations of `launch`, `build`, and `cache` booleans are valid. When a layer declares more than one type (e.g. `launch = true` and `cache = true`), the requirements of each type apply.

> `build`   | `cache`  | `launch` | Metadata Restored        | Layer Restored      

Reviewer: ```suggestion
The following table illustrates the behavior depending on the value of each flag:
`build`   | `cache`  | `launch` | Metadata Restored        | Layer Restored      
```

Author: Please see https://github.com/buildpacks/spec/pull/209/commits/35eda43c391bcbbdd15fc7858ed9ec7f9e865f3d.

> Note that the lifecycle only restores layers from the cache due to performance reasons.

Author: Is it correct to write the following:
`Note that the lifecycle only restores layers from the cache, never from the previous image, due to performance reasons.`
?

> Note that the lifecycle only restores layers from the cache due to performance reasons.

Author: Left the "performance" part outside, since we don't use to write such things in the spec.
Please see https://github.com/buildpacks/spec/pull/209/commits/3cd0f8be71cf90f563cbaa788ad1758f7d00f3ad.

`build`   | `cache`  | `launch` | Metadata Restored        | Layer Restored      
----------|----------|----------|--------------------------|---------------------

> T         | T        | T        | Yes - from the app image | Yes* - from the cache

Author: You're right. I'll make this change!

> T         | T        | T        | Yes - from the app image | Yes* - from the cache

Author: Please see https://github.com/buildpacks/spec/pull/209/commits/733f8b4537c89b99affec180e6a78fd4822fb73e.

F         | T        | T        | Yes - from the app image | Yes - from the cache
F         | T        | F        | Yes - from the cache     | Yes - from the cache
F         | F        | T        | Yes - from the app image | No
F         | F        | F        | No                       | No

Notes:

> * The metadata and layer are restored only if they match each other (the cache can be changed/deleted).

Reviewer: ```suggestion
* The metadata and layer are restored only if they match each other.
```
For some reason I find the extra info confusing, but maybe that's just me :/

Author: Personally, I think it explains how the metadata and layer can be different.
But I don't have any strong opinion.

Reviewer 2: Hi! I don't understand this sentence, even prior to removing the extra info. By "match each other" I'm presuming it means "the metadata matches the layer contents", but I don't see how the Lifecycle would know whether the two match? For example if the metadata describes the layer as containing a particular version of Node, and the layer contents contains a different version, only the buildpack would have the domain specific knowledge to make that determination, not the Lifecycle.

Which makes me believe this sentence must actually be meaning something else?

Author: Hi @edmorley, thank you for your feedback!
I rephrased this sentence. You can see it here: https://github.com/buildpacks/spec/pull/209/commits/3077962a6086ef53124fe502437891b6b90b2049.
Please let me know if you still have any questions.

Reviewer 2: @yaelharel with that edit I understand now - thank you! :-)

\* The metadata and layer are restored only if the layer SHA recorded in the previous image matches the layer SHA recorded in the cache.

> Examples:

Reviewer 3: I like the examples here! We historically haven't done examples in the spec. @sclevine @ekcasey is this something we want to keep here or move somewhere else (in that reference doc that we keep talking about).

Author: Hi @hone, I was also wondering on whether this is the right place for it (although it felt right :) ) so we had [this](https://buildpacks.slack.com/archives/CCP60GJLS/p1616444461016800) conversation over slack and @sclevine and @ekcasey agreed that it's a good idea to add the examples here.

Reviewer 3: hi @yaelharel. I'm very 👍 for examples fwiw, I didn't see that slack convo. It does make the rest of the spec look really lame in comparison. ;) I'll mark this as resolved.

> If the version of the jdk in the cache is the one that is needed, a downstream buildpack will use it.  

Reviewer: ```suggestion
Example: a Java buildpack might read the restored layer metadata to determine if the version of the JDK used in the previous build of the OCI image is the one that is needed. If so, it might choose to re-use the layer from the cache to avoid re-downloading.
```

Author: Please see https://github.com/buildpacks/spec/pull/209/commits/35eda43c391bcbbdd15fc7858ed9ec7f9e865f3d.

> * `build == T, cache == T, launch == F`:  

Reviewer: ```suggestion
* `build == true, cache == true, launch == false`:  
```

Author: Please see https://github.com/buildpacks/spec/pull/209/commits/3cd0f8be71cf90f563cbaa788ad1758f7d00f3ad.

* `build == T, cache == F, launch == T`: Ruby buildback.  

> A downstream buildpack needs it in order to start the app, but the cache isn't used for some reason.
> There is no reason to restore only the metadata since the layer is needed (and layers aren't restored from the app image).

Reviewer: ```suggestion
Example: a Ruby buildpack might need to provide Ruby to a downstream buildpack (such as bundler) and also include Ruby in the exported OCI image so that it could be used to start the app at runtime. If Ruby is not cached, there is no reason to restore only the metadata as the layer is needed at build time (and layers aren't restored from the previous app image).
```

Author: Looks good. I'll add to the end: `(and layers aren't restored from the previous image)`.

Author: Please see https://github.com/buildpacks/spec/pull/209/commits/35eda43c391bcbbdd15fc7858ed9ec7f9e865f3d.

> * `build == T, cache == F, launch == T`:  

Reviewer: ```suggestion
* `build = true, cache = false, launch = true`:  
```

Author: Please see https://github.com/buildpacks/spec/pull/209/commits/3cd0f8be71cf90f563cbaa788ad1758f7d00f3ad.

> * `build == T, cache == F, launch == F`: CA-certificates buildback.

Reviewer: Are there other ca-certificates buildpacks besides the Paketo one? We might want to be a bit more generic so that we stay vendor-neutral.

Author: We changed it as part of https://github.com/buildpacks/spec/pull/209/commits/35eda43c391bcbbdd15fc7858ed9ec7f9e865f3d, so I'm going to mark this conversation as resolved.

> Downstream buildpacks need it, but don't ever want to store it for next builds (always read it from the bindings).

Reviewer: ```suggestion
Example: a buildpack that reads from a bind-mounted directory at build time and provides data to downstream buildpacks.
```

Author: Please see https://github.com/buildpacks/spec/pull/209/commits/35eda43c391bcbbdd15fc7858ed9ec7f9e865f3d.

* `build == F, cache == F, launch == T`: Java buildpack.  

> If the version of the jre in the previous app image is the one that is needed, there is no need to download it again.

Reviewer: ```suggestion
Example: a Java buildpack might read the restored layer metadata to determine if the version of the JRE included in the previous build of the OCI image is the one that is needed. If so, it might choose to re-use the layer from the previous image to avoid re-downloading.
```

Author: Please see https://github.com/buildpacks/spec/pull/209/commits/35eda43c391bcbbdd15fc7858ed9ec7f9e865f3d.

> A Ruby buildpack might need to provide Ruby to a downstream buildpack (such as bundler) and also include Ruby in the exported OCI image so that it could be used to start the app at runtime. If the Ruby layer is not cached, the layer metadata will not be restored as the Ruby buildpack will always need to re-download Ruby (layers are not restored from the previous image).

Reviewer 4: Maybe this is a bad example, because most buildpack authors in this case would want to cache ruby to avoid time spent redownloading? Another example might be a layer that contains environment variables needed during both build and launch because restoring these from the cache doesn't necessarily save time.

Reviewer 3: Is this too much for the spec?

Author: @hone, following a conversation that I had with @ekcasey and @natalieparellano, we decided to change this example to a "all true" example. Please see https://github.com/buildpacks/spec/pull/209/commits/3cd0f8be71cf90f563cbaa788ad1758f7d00f3ad.

> * `build == T, cache == F, launch == F`:  

Reviewer: ```suggestion
* `build == true, cache == false, launch == false`:  
```

Author: Please see https://github.com/buildpacks/spec/pull/209/commits/3cd0f8be71cf90f563cbaa788ad1758f7d00f3ad.

> * `build == F, cache == F, launch == T`:  

Reviewer: ```suggestion
* `build == false, cache == false, launch == true`:  
```

Author: Please see https://github.com/buildpacks/spec/pull/209/commits/3cd0f8be71cf90f563cbaa788ad1758f7d00f3ad.

T         | T        | T        | Yes - from the app image | Yes - from the cache
T         | T        | F        | Yes - from the cache     | Yes - from the cache
T         | F        | T        | No                       | No
T         | F        | F        | No                       | No
F         | T        | T        | Yes - from the app image | Yes - from the cache
F         | T        | F        | Yes - from the cache     | Yes - from the cache
F         | F        | T        | Yes - from the app image | No
F         | F        | F        | No                       | No

Notes:
* The metadata and layer are restored only if they match each other (the cache can be changed/deleted).
* The layer is restored only from the cache due to performance reasons.

Examples:
* `build == T, cache == T, launch == F`:  
A Java buildpack might read the restored layer metadata to determine if the version of the JDK used in the previous build of the OCI image is the one that is needed. If so, it might choose to re-use the layer from the cache to avoid re-downloading the JDK.
* `build == T, cache == F, launch == T`:  

> A Ruby buildpack might need to provide Ruby to a downstream buildpack (such as bundler) and also include Ruby in the exported OCI image so that it could be used to start the app at runtime. If the Ruby layer is not cached, the layer metadata will not be restored as the Ruby buildpack will always need to re-download Ruby (and layers are not restored from the previous image).

Reviewer: ```suggestion
A Ruby buildpack might need to provide Ruby to a downstream buildpack (such as bundler) and also include Ruby in the exported OCI image so that it could be used to start the app at runtime. If the Ruby layer is not cached, the layer metadata will not be restored as the Ruby buildpack will always need to re-download Ruby as layers are not restored from the previous image.
```

Author: Please see https://github.com/buildpacks/spec/pull/209/commits/36fc622b710d41389a76422c0f6405b3a744b6d1.

> Using the [Layer Content Metadata](#layer-content-metadata-toml) provided by a buildpack in a `<layers>/<layer>.toml` file, the lifecycle MUST determine:

Reviewer 5: ```suggestion
Using the [Layer Content Metadata](#layer-content-metadata-toml) provided by a buildpack in the `[types]` table of a `<layers>/<layer>.toml` file, the lifecycle MUST determine:
```

Author: Please see https://github.com/buildpacks/spec/pull/209/commits/582d68443291d2a6fe2f12d463dd33ef44c65eea.

* `build = false, cache = false, launch = true`:
A Java buildpack might read the restored layer metadata to determine if the version of the JRE included in the previous build of the OCI image is the one that is needed. If so, it might choose to re-use the layer from the previous image to avoid re-downloading the JRE.

#### Launch Layers

A buildpack MAY specify that a `<layers>/<layer>/` directory is a launch layer by placing `launch = true` under `[types]` in `<layers>/<layer>.toml`.

The lifecycle MUST make all launch layers accessible to the app as described in the [Environment](#environment) section.

The lifecycle MUST include each launch layer in the built OCI image.
The lifecycle MUST also store the Layer Content Metadata associated with each layer so that it can be recovered using the layer Diff ID.

Before a given re-build:
- If a launch layer is marked `cache = false` and `build = false` in the previous image metadata, the lifecycle:
  - MUST restore Layer Content Metadata to `<layers>/<layer>.toml`, excluding the `launch`, `build`, and `cache` booleans.
  - MUST NOT restore the corresponding `<layers>/<layer>/` directory from any previous build.

After a given re-build:

> - If a buildpack keeps `launch = true` under `[types]` in `<layers>/<layer>.toml` and leaves no `<layers>/<layer>/` directory, the lifecycle:

Author: You're right. I missed it.

> - If a buildpack keeps `launch = true` under `[types]` in `<layers>/<layer>.toml` and leaves no `<layers>/<layer>/` directory, the lifecycle:

Author: I changed it yesterday. Please see https://github.com/buildpacks/spec/pull/209/commits/c988be3768b4357d9a0b7a349aa8e7c9b64124f9.

> - If a buildpack keeps `launch = true` under `[types]` in `<layers>/<layer>.toml` and leaves no `<layers>/<layer>/` directory, the lifecycle:

Reviewer 4: sorry I think this was a stale comment I failed to successfully submit yesterday

A buildpack MAY specify that a `<layers>/<layer>/` directory is a cached layer by placing `cache = true` in `<layers>/<layer>.toml`.

If a cache is provided, the lifecycle:
- SHOULD store all cached layers after a successful build.
- SHOULD store the Layer Content Metadata associated with each layer so that it can be recovered using the layer Diff ID

Before the next re-build:
- The lifecycle MUST do both or neither of the following:
  - Restore Layer Content Metadata to `<layers>/<layer>.toml`, excluding the `launch`, `build`, and `cache` booleans.
  - Restore layer contents to the `<layers>/<layer>/` directory.

#### Ignored Layers

Layers marked `launch = false`, `build = false`, and `cache = false` behave like temporary directories, available only to the authoring buildpack, existing for the duration of a single build.

At the end of each individual buildpack's build phase:
- The lifecycle:
  - MUST rename `<layers>/<layer>/` to `<layers>/<layer>.ignore/` for all layers where `launch = false`, `build = false`, and `cache = false`, in order to prevent subsequent buildpacks from accidentally depending on an ignored layer.

> The lifecycle restores all layers from previous builds as ignored layers. Buildpacks MUST set `launch`, `build`, and/or `cache` in `<layers>/<layer>.toml` to use a restored layer.

Reviewer 3: Is an "ignored" layer one that has `build`, `launch`, and `cache` set to false?

Reviewer 4: Yes. The first line in the "gnored layers" sections attempts to communicate that.

Author: @hone, I agree that this sentence might be a bit confusing.
@ekcasey, the lifecycle restores all layers from previous builds and sets the 3 flags to false, but it doesn't rename `<layers>/<layer>/` to `<layers>/<layer>.ignore/`. So maybe it's not very accurate to write that the layers are restored as `ignored layers`.
What do you think?

Reviewer: This paragraph seems pretty well covered in the `Reusing Layers`. I feel like we could just remove it.

Author: After talking with @ekcasey and @natalieparellano, we decided to delete this line.
Please see https://github.com/buildpacks/spec/pull/209/commits/3cd0f8be71cf90f563cbaa788ad1758f7d00f3ad.

Before the next re-build:
- The lifecycle MUST do both or neither of the following:
  - Restore Layer Content Metadata to `<layers>/<layer>.toml`, excluding the `[types]` table.
  - Restore layer contents to the `<layers>/<layer>/` directory.

#### Ignored Layers

> Layers marked `launch = false`, `build = false`, and `cache = false` (or when the `[types]` table does not exist) behave like temporary directories, available only to the authoring buildpack, existing for the duration of a single build.

Author: I'm not sure if we need to add this parenthesis. The same thoughts are on line 278.

Author: Changed as part of https://github.com/buildpacks/spec/pull/209/commits/f2eb2af7d05dff04d6570e2057321c45adca96ac..

> Layers marked `launch = false`, `build = false`, and `cache = false` (or when the `[types]` table does not exist) behave like temporary directories, available only to the authoring buildpack, existing for the duration of a single build.

Reviewer 4: Maybe we can remove this parenthetical and add a someone near [here](https://github.com/buildpacks/spec/pull/209/commits/6f968784d15143d02588a0b2dbbd5f1c92ce1dbd#diff-d5ffdcac80bcacec34b3a8a1ba9da8efbbc4df5ad118ec581201736330f73183R199) along the lines of: "The lifecycle MUST treat a layer with unset `types` as a `launch = false`, `build = false`, `cache = false` layer."

Author: Please see https://github.com/buildpacks/spec/pull/209/commits/f2eb2af7d05dff04d6570e2057321c45adca96ac.

During the build phase, typical buildpacks might:

1. Read the Buildpack Plan in `<plan>` to determine what dependencies to provide.
1. Provide the application with dependencies for launch in `<layers>/<layer>`.

> 1. Reuse application dependencies from a previous image by appending by `launch = true` to `<layers>/<layer>.toml`.

Reviewer: ```suggestion
1. Reuse application dependencies from a previous image by appending `types.launch = true` to `<layers>/<layer>.toml`.
```

Author: @natalieparellano, I thought that you, @ekcasey and I decided to keep using `launch`, `build` and `cache` and not change them to `types.launch`, `types.build` and `types.cache` for clarity purposes.
If we decide to add `types`, there are more places in this file that should be changed.

Reviewer: Summarizing side conversation - we think that when talking about "layers" marked `cache = true` we don't need `types`, but we should give accurate instructions for modifying the file.

Reviewer 5: I think we should add a sentence or two at the beginning of the `Layer Types` section that explains these keys are under the `[types]` table, and then we can refer to them only as `launch`, `build`, `cache` from there on (but only within the `Layer Types` section. If there are uses outside of that section, we should reevaluate.

Author: Please see https://github.com/buildpacks/spec/pull/209/commits/81cb981a51b6186238559a28a327366160fa2024.

> 1. Reuse application dependencies from a previous image by appending `[types]` and `launch = true` to `<layers>/<layer>.toml`.

Reviewer: ```suggestion
1. Reuse application dependencies from a previous image by appending `[types]`, `launch = true` to `<layers>/<layer>.toml`.
```

1. Provide subsequent buildpacks with dependencies in `<layers>/<layer>`.

> 1. Reuse cached build dependencies from a previous build by appending `build = true` and `cache = true` to `<layers>/<layer>.toml`.

Reviewer: ```suggestion
1. Reuse cached build dependencies from a previous build by appending `types.build = true` and `types.cache = true` to `<layers>/<layer>.toml`.
```

Author: Please see https://github.com/buildpacks/spec/pull/209/files#r600947906.

Author: Please see https://github.com/buildpacks/spec/pull/209/commits/81cb981a51b6186238559a28a327366160fa2024.

   - The executable component of the lifecycle that implements the launch phase, and
   - An `ENTRYPOINT` set to that component.

> Finally, any `<layers>/<layer>` directories specify `cache = true` under `[types]` in `<layers>/<layer>.toml` MAY be preserved for the next local build.

Reviewer: ```suggestion
Finally, any `<layers>/<layer>` directories specified as `cache = true` under `[types]` in `<layers>/<layer>.toml` MAY be preserved for the next local build.
```

Author: Please see https://github.com/buildpacks/spec/pull/209/commits/582d68443291d2a6fe2f12d463dd33ef44c65eea.

### Layer Content Metadata (TOML)

```toml

> [types]

Reviewer 5: Was surprised to see this. I don't think we need to use `types.launch` in every mention, but the `Layer Types` section should make it clear that the keys are nested under this table. That, or we just don't introduce this table in this PR.

Author: Hi @jkutner, thank you for your feedback!

I've just added a commit that specifying that these flags are under `[types]` every time that we refer to `<layers>/<layer>.toml`. Please see https://github.com/buildpacks/spec/pull/209/commits/81cb981a51b6186238559a28a327366160fa2024.

At the beginning of the `Layer Types` section, we have a link to `Layer Content Metadata`, where we specify that all of the flags are under `[types]`. Do you think we should make it clearer than that?
I'm afraid that we want to specify it now only because it's new, but actually it's not necessary.

```toml
[types]

>   launch = true

Author: I changed all of the flags to true because I feel that it's clearer this way since we omit the table when they're set to false.

Author: Please see https://github.com/buildpacks/spec/pull/209/commits/f2eb2af7d05dff04d6570e2057321c45adca96ac.
